In [1]:
import sys
import os
sys.path.append(os.path.abspath('../scripts'))

# Get the current directory
current_dir = os.getcwd()
print(current_dir)

# Get the parent directory
parent_dir = os.path.dirname(current_dir)
print(parent_dir)

# Insert the path to the parent directory
sys.path.insert(0, parent_dir)

from load_data import load_data_using_sqlalchemy

c:\Users\HP\Documents\10-Academy\Telecom-WEEK2\notebooks
c:\Users\HP\Documents\10-Academy\Telecom-WEEK2


database connection and loading data

In [2]:
# Define your SQL query
query = "SELECT * FROM xdr_data;"  # Replace with your actual table name

# Load data from PostgreSQL using SQLAlchemy
df = load_data_using_sqlalchemy(query)

# Display the first few rows of the dataframe
if df is not None:
    print("Successfully loaded the data")
else:
    print("Failed to load data.")

Successfully loaded the data


In [3]:
from scripts.user_overview_analysis import OverviewAnalyzer


In [5]:
analyzer = OverviewAnalyzer(df, 'HTTP DL (Bytes)')

In [6]:
missing_values_df = analyzer.missing_values_table(df)
print("Missing Values in df:")
print(missing_values_df)

Your selected dataframe has 55 columns.
There are 41 columns that have missing values.
Missing Values in df:
                                          Missing Values  % of Total Values  \
Nb of sec with 37500B < Vol UL                    130254               86.8   
Nb of sec with 6250B < Vol UL < 37500B            111843               74.6   
Nb of sec with 125000B < Vol DL                    97538               65.0   
TCP UL Retrans. Vol (Bytes)                        96649               64.4   
Nb of sec with 31250B < Vol DL < 125000B           93586               62.4   
Nb of sec with 1250B < Vol UL < 6250B              92894               61.9   
Nb of sec with 6250B < Vol DL < 31250B             88317               58.9   
TCP DL Retrans. Vol (Bytes)                        88146               58.8   
HTTP UL (Bytes)                                    81810               54.5   
HTTP DL (Bytes)                                    81474               54.3   
Avg RTT DL (ms)       

Analysis of Missing Values in the Telecom Dataset
The dataset contains 55 columns, of which 41 columns have missing values. This high proportion of missing data across numerous features is a significant concern, as it can impact the quality and reliability of the analysis. Here's a breakdown of the findings:

Key Findings:
Significant Missing Data in High-Volume Features:

Columns like Nb of sec with 37500B < Vol UL, Nb of sec with 6250B < Vol UL < 37500B, and Nb of sec with 125000B < Vol DL have substantial missing values, with percentages of missing data as high as 86.8%, 74.6%, and 65.0%, respectively.
These columns are critical as they represent key performance indicators related to data upload and download volumes. The high percentage of missing data suggests either data collection issues or irregularities in user activity.
Transmission and HTTP Data:

Columns related to TCP retransmission and HTTP data, such as TCP UL Retrans. Vol (Bytes) and HTTP UL (Bytes), also exhibit a high percentage of missing values (around 54-64%).
Missing data in these features might hinder the understanding of network performance and user behavior, especially in analyzing retransmission rates and HTTP traffic.
Moderate Missing Data in Handset Information:

The Handset Manufacturer, Handset Type, and IMEI columns have around 0.4% missing data. Although the percentage is relatively low, it is crucial to address these missing values as they are fundamental to the User Overview analysis, which focuses on handset types and manufacturers.
Accurate and complete handset information is essential for identifying trends in device usage and making informed marketing recommendations.
Critical Identifiers and Time Features:

Columns such as MSISDN/Number, Bearer Id, Start, End, and Dur. (ms) show minimal missing values (less than 1%). These columns are vital for user identification, session tracking, and duration analysis.
Ensuring these features are complete and accurate is critical for analyzing user behavior and session durations.
Other Technical Metrics:

Several columns, such as UL TP < 10 Kbps (%), DL TP > 1 Mbps (%), and Total UL (Bytes), have very minimal missing values (less than 1%).
These columns provide detailed insights into data transmission rates and overall data usage, which are crucial for performance analysis.
Interpretation:
The dataset has significant missing values in key performance and volume-related features, which could potentially skew the analysis if not properly addressed. While some features have minimal missing data, the overall high percentage of missing data across the dataset indicates potential data collection issues, user behavior irregularities, or device-related problems.

Recommendations:
Data Imputation or Exclusion: Consider imputation techniques for missing values where appropriate, especially for columns with a smaller percentage of missing data. For features with extremely high missing values (over 50%), assess the impact on the analysis and consider excluding these columns if they do not add substantial value.

Focus on Handset Data: Ensure that the handset information is as complete as possible since it is crucial for the User Overview analysis. Missing data in these columns should be addressed with targeted imputation strategies or by investigating the source of the missing data.

Investigate Data Collection: The high level of missing data in certain columns may indicate issues with the data collection process. Investigating the root causes of these missing values could help improve data quality in future collections.

Sensitivity Analysis: Conduct sensitivity analysis to understand how the missing data might impact the results. This can help in determining the robustness of your findings and ensuring that the analysis remains reliable even with incomplete data.

This analysis provides an overview of the dataset’s condition and highlights areas where improvements or careful consideration are necessary for a more accurate and insightful analysis.

The telecom dataset I'm working with contains 55 columns, and I've found that 41 of these columns have missing values. This is concerning, especially since key features related to data upload and download volumes, like Nb of sec with 37500B < Vol UL and Nb of sec with 125000B < Vol DL, have missing values ranging from 65% to nearly 87%. The high percentage of missing data in these areas could indicate problems with data collection or irregularities in user activity. I've also noticed that important columns related to TCP retransmission and HTTP data, such as TCP UL Retrans. Vol (Bytes) and HTTP UL (Bytes), have missing values of around 54-64%. This could affect my analysis of network performance and user behavior. On a more positive note, the handset-related data, such as Handset Manufacturer and Handset Type, has a much lower percentage of missing values, around 0.4%. However, even this small amount of missing data needs to be addressed since it's crucial for my analysis of device usage. Columns that are vital for tracking user sessions and behavior, like MSISDN/Number and Bearer Id, have minimal missing values, which is reassuring. Given these findings, I'm considering using imputation techniques for columns with lower percentages of missing data, while possibly excluding those with extremely high missing values. I also think it's important to investigate the causes of these missing values to improve data quality in future projects. Overall, I'm aware that these missing values could impact my analysis, so I'll be conducting sensitivity checks to ensure that my findings are as accurate and reliable as possible.

In [8]:
df['x'] = analyzer.convert_bytes_to_megabytes(df, 'HTTP DL (Bytes)')

print(df)

TypeError: OverviewAnalyzer.convert_bytes_to_megabytes() takes 2 positional arguments but 3 were given